rusk/src/lib/chain/rusk.rs

```diff
@@ -495,15 +506,58 @@ fn execute(
     session: &mut Session,
     tx: &PhoenixTransaction,
 ) -> Result<CallReceipt<Result<Vec<u8>, ContractError>>, PiecrustError> {
+    if let Some(deploy) = tx.payload().contract_deploy() {
+        let bytecode_charge =
+            deploy.bytecode.bytes.len() as u64 * GAS_PER_DEPLOY_BYTE;
+        if tx.payload().fee.gas_limit < bytecode_charge {
+            return Err(PiecrustError::Panic(
+                "not enough gas to deploy".into(),
+            ));
+        }
+    }
+
+    let tx_stripped = strip_off_bytecode(tx);
     // Spend the inputs and execute the call. If this errors the transaction is
     // unspendable.
     let mut receipt = session.call::<_, Result<Vec<u8>, ContractError>>(
         TRANSFER_CONTRACT,
         "spend_and_execute",
-        tx,
+        tx_stripped.as_ref().unwrap_or(tx),
         tx.payload().fee.gas_limit,
     )?;
 
+    // Deploy if this is a deployment transaction
+    if let Some(deploy) = tx.payload().contract_deploy() {
```

Comment:
I *think* I understand the code underneath, but it would be nice to have it commented as per its intent. In particular, I would like to understand under which conditions this merits re-execution/not.

Reply:
Added comments explaining the deployment code in more detail including the re-execution aspect of deployment transactions.